otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py

```diff
@@ -171,3 +191,202 @@ def _balance_losses(self, losses_cls, losses_reg):
+        )
+        self.cur_loss_idx += 1
+        return losses
+
+    def _store_loss_dyns(self, losses: torch.Tensor, key: TrackingLossType) -> None:
+        loss_dyns = self.loss_dyns[key]
+        for batch_idx, bbox_idx, loss_item in zip(self.batch_inds, self.bbox_inds, losses.detach().cpu()):
+            loss_dyns[(batch_idx.item(), bbox_idx.item())].add(loss_item.item())
+
+    def _get_pos_inds(self, labels):
+        pos_inds = super()._get_pos_inds(labels)
+
+        if len(pos_inds) > 0:
+            pos_assigned_gt_inds = self.all_pos_assigned_gt_inds[self.cur_loss_idx].reshape(-1)
+
+            gt_inds = pos_assigned_gt_inds[pos_inds].cpu()
+
+            self.batch_inds = gt_inds // self.max_gt_bboxes_len
+            self.bbox_inds = gt_inds % self.max_gt_bboxes_len
+
+        self.pos_inds = pos_inds
+        return pos_inds
+
+    def _get_loss_cls(self, num_total_samples, loss_cls_all, pos_inds, topk_loss_cls_neg):
+        loss_cls_pos = loss_cls_all[pos_inds]
+        loss_cls_neg = topk_loss_cls_neg.sum()
+        loss_cls = (loss_cls_pos.sum() + loss_cls_neg) / num_total_samples
+
+        self._store_loss_dyns(loss_cls_pos.detach(), TrackingLossType.cls)
+        return loss_cls
+
+    def _get_loss_bbox(self, bbox_pred, bbox_targets, bbox_weights, num_total_samples):
+        loss_bbox = smooth_l1_loss(
+            bbox_pred,
+            bbox_targets,
+            bbox_weights,
+            beta=self.train_cfg.smoothl1_beta,
+            avg_factor=num_total_samples,
+            reduction="none",
+        )
+
+        self._store_loss_dyns(loss_bbox[self.pos_inds].detach().mean(-1), TrackingLossType.bbox)
+        return weight_reduce_loss(loss_bbox, reduction="mean", avg_factor=num_total_samples)
+
+    def get_targets(
+        self,
+        anchor_list,
+        valid_flag_list,
+        gt_bboxes_list,
+        img_metas,
+        gt_bboxes_ignore_list=None,
+        gt_labels_list=None,
+        label_channels=1,
+        unmap_outputs=True,
+        return_sampling_results=False,
+    ):
+        """Get targets for Detection head."""
+        num_level_anchors = [anchors.size(0) for anchors in anchor_list[0]]
+        self.max_gt_bboxes_len = max([len(gt_bboxes) for gt_bboxes in gt_bboxes_list])
+        self.cur_batch_idx = 0
+        self.pos_assigned_gt_inds_list = []
+        targets = super().get_targets(
+            anchor_list,
+            valid_flag_list,
+            gt_bboxes_list,
+            img_metas,
+            gt_bboxes_ignore_list,
+            gt_labels_list,
+            label_channels,
+            unmap_outputs,
+            return_sampling_results,
+        )
+
+        all_pos_assigned_gt_inds = images_to_levels(self.pos_assigned_gt_inds_list, num_level_anchors)
+        self.all_pos_assigned_gt_inds = torch.cat(all_pos_assigned_gt_inds, -1)
+        return targets
+
+    def _get_targets_single(
```

Comment:
why this should be required?

Reply:
This is because we should store `sampling_result.pos_assigned_gt_inds` to match the loss dynamics statistics obtained from the anchors with the source bounding boxes.
https://github.com/openvinotoolkit/training_extensions/blob/cbf7ecc05780fc23d5d854a59942beb0d401f539/otx/algorithms/detection/adapters/mmdet/models/heads/custom_ssd_head.py#L347-L359